pkg/controller/usersignup/approval.go

```diff
@@ -0,0 +1,120 @@
+	config := &toolchainv1alpha1.HostOperatorConfig{}
+	if err := cl.Get(context.TODO(), types.NamespacedName{Namespace: userSignup.Namespace, Name: "config"}, config); err != nil {
+		if apierrors.IsNotFound(err) {
+			if userSignup.Spec.Approved {
+				return true, getOptimalTargetCluster(userSignup, getMemberClusters), nil
+			}
+			return false, "", nil
+		}
+		return false, "", errors.Wrapf(err, "unable to read HostOperatorConfig resource")
+	}
+	if !userSignup.Spec.Approved && !config.Spec.AutomaticApproval.Enabled {
+		return false, "", nil
+	}
+
+	status := &toolchainv1alpha1.ToolchainStatus{}
+	if err := cl.Get(context.TODO(), types.NamespacedName{Namespace: userSignup.Namespace, Name: crtConfig.GetToolchainStatusName()}, status); err != nil {
+		return false, "", errors.Wrapf(err, "unable to read ToolchainStatus resource")
+	}
+	counts, err := counter.GetCounts()
+	if err != nil {
+		return false, "", errors.Wrapf(err, "unable to get the number of provisioned users")
+	}
+
+	targetCluster := getOptimalTargetCluster(userSignup, getMemberClusters, hasNotReachedMaxNumberOfUsersThreshold(config, counts), hasEnoughResources(config, status))
+	if targetCluster == "" {
+		return false, "", fmt.Errorf("no suitable member cluster found")
+	}
+
+	// approved manually
+	if userSignup.Spec.Approved {
+		return true, targetCluster, nil
+	}
+
+	if config.Spec.AutomaticApproval.MaxNumberOfUsers.Overall != 0 {
+
+		if config.Spec.AutomaticApproval.MaxNumberOfUsers.Overall <= counts.MasterUserRecordCount {
+			return false, "", nil
+		}
+	}
+
+	if config.Spec.AutomaticApproval.ResourceCapacityThreshold.DefaultThreshold != 0 {
+		if config.Spec.AutomaticApproval.MaxNumberOfUsers.Overall <= counts.MasterUserRecordCount {
+			return false, "", nil
+		}
+	}
+	return true, targetCluster, nil
+}
+
+func hasNotReachedMaxNumberOfUsersThreshold(config *toolchainv1alpha1.HostOperatorConfig, counts counter.Counts) cluster.Condition {
+	return func(cluster *cluster.CachedToolchainCluster) bool {
+		numberOfUserAccounts := counts.UserAccountsPerClusterCounts[cluster.Name]
+		threshold := config.Spec.AutomaticApproval.MaxNumberOfUsers.SpecificPerMemberCluster[cluster.Name]
+		return threshold == 0 || numberOfUserAccounts <= threshold
+	}
+}
+
+func hasEnoughResources(config *toolchainv1alpha1.HostOperatorConfig, status *toolchainv1alpha1.ToolchainStatus) cluster.Condition {
+	return func(cluster *cluster.CachedToolchainCluster) bool {
+		threshold, found := config.Spec.AutomaticApproval.ResourceCapacityThreshold.SpecificPerMemberCluster[cluster.Name]
+		if !found {
+			threshold = config.Spec.AutomaticApproval.ResourceCapacityThreshold.DefaultThreshold
+		}
+		if threshold == 0 {
+			return true
+		}
+		for _, memberStatus := range status.Status.Members {
+			if memberStatus.ClusterName == cluster.Name {
```

Comment:
could we refactor this to avoid some of the nesting by any chance?

Reply:
hmmm... do you have any proposal?

Reply:
How about if we combine the first two if statements?

```suggestion
			if memberStatus.ClusterName == cluster.Name && len(memberStatus.MemberStatus.ResourceUsage.MemoryUsagePerNodeRole) > 0 {
```

And then move the following to a separate function named `clusterHasEnoughResources` or something?
```
for _, usagePerNode := range memberStatus.MemberStatus.ResourceUsage.MemoryUsagePerNodeRole {
	if usagePerNode >= threshold {
		return false
	}
}
return true
```

I guess that would help with readability?

Reply:
> How about if we combine the first two if statements?

I technically could, but it would go through all member statuses no matter if the correct one was already found or not (for the case when the memory usage is `0`). I know that we won't probably have many of member clusters with such number, but I would like to stay with the proper solution if you don't mind.
In addition, I'm not a fan of such a long if statements - for me it's harder to read.

If it helps, I can separate the inner part - see: 8462fab
is it better @tinakurian @rajivnathan

Reply:
Looks good for me. Thanks for adjusting it Matous!

Reply:
looks good @MatousJobanek!